### PR TITLE
CSHARP-2705: Linq Translation of Nested Any throws NotSupportedException.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
Evergreen: https://evergreen.mongodb.com/version/5d91f3362fbabe5e806b9da2

### CODE REVIEW

**Comment by DmitryLukyanov** [comment]
this ticket contains additional cases related to this test: https://github.com/mongodb/mongo-csharp-driver/blob/master/tests/MongoDB.Driver.Tests/Linq/Translators/PredicateTranslatorValidationTests.cs#L52 from the ticket CSHARP-2460

File: tests/MongoDB.Driver.Tests/Linq/Translators/PredicateTranslatorValidationTests.cs [hunk 1]
```diff
@@ -59,6 +59,75 @@ public void Should_not_throw_the_exception_when_a_child_expression_is_called_fro
...
+        {
+            Setup();
+
+            var local = new List<int> { 1, 2 };
+            var local2 = new List<bool> { true, false };
+
+            Expression<Func<TestObject, bool>> expr = (a) => a.Collection1.Any(b => local.Any(c => b.Collection2.Contains(c)));
+            AssertWhere(expr, "{ $match : { 'Collection2' : { '$in' : [1, 2] } } }");
+
+            expr = (a) => a.Collection1.Any(b => b.Collection1.Any(c => local.Any(d => c.Collection2.Contains(d))));
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/Linq/Translators/PredicateTranslatorValidationTests.cs:73]: This linq query generates the same mongo query as in the line above which is strange. It doesn't look like the behavior related to this ticket, but I would create a new ticket to investigate the reasons of it.

File: src/MongoDB.Driver/Linq/Processors/OutOfCurrentScopePrefixCollector.cs [hunk 2]
```diff
@@ -41,12 +42,25 @@ public static IEnumerable<string> Collect(Expression expression)
...
             }
 
             return base.Visit(node);
         }
+
+        protected internal override Expression VisitWhere(WhereExpression node)
+        {
+            if (node.Source.NodeType == ExpressionType.Constant)
+            {
+                _isParentConstantExpression = true;
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/Linq/Processors/OutOfCurrentScopePrefixCollector.cs]: Now, we have a flag about whether a source of a `where` expression is a `constantExpression` or not.

Comment (rstam) [reply src/MongoDB.Driver/Linq/Processors/OutOfCurrentScopePrefixCollector.cs]: Does the name of this field need to be more accurate?

Perhaps: `_isWhereSourceConstantExpression`?

`_isParentConstantExpression` is misleading.

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver/Linq/Processors/OutOfCurrentScopePrefixCollector.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Processors/OutOfCurrentScopePrefixCollector.cs]: done

**Comment by DmitryLukyanov** [comment]
ready for rereview.

**Review by rstam — APPROVED** [review]

**Pull request merged**